packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -821,27 +832,6 @@ export default function makeKernelKeeper(
     insistKernelType('promise', kernelSlot);
     insistMessage(msg);
 
-    // Each message on a promise's queue maintains a refcount to the promise
```

Comment:
Ok, so messages queued to an object retain a refcount to their target object, but messages queued *on* a promise do not. Seems reasonable, given the container-ish relationship between promise, queue, and message.

Reply:
Ugh? No they do definitely keep a refcount to the target promise. What makes you say that?

Reply:
you deleted the portion of `addMessageToPromiseQueue` that increments the refcounts on everything in the message

(this comment is supposed to be associated with the deleted lines 825-845 of `kernelKeeper.js`.. I'm not sure that GitHub renders comments correctly when attached to deleted lines, vs added lines)

yeah now I'm not sure why I said "seems reasonable".. are these refcounts being incremented somewhere else instead now?

Reply:
Ha now I understand where the comment comes from.

So with the new structure, a ref count is added when messages are first added to the acceptance queue right after the syscall is received, and decremented when removed from the run-queue just before the delivery. When moved between queues the refcount doesn't change, except when retargeting from a promise to an object, which currently can only happen when [moving from the acceptance queue to the run queue](https://github.com/Agoric/agoric-sdk/pull/5252/files#diff-93cec0b19f5deab7c2117e72ee0a19eec0957fc7f6ea0445a8c8f6d3e5a7684fR1107-R1108). In the next PR retargeting will also possibly happen when moving from the promise queue to the run-queue (currently it goes back to the acceptance queue) when resolving / setting the decider.

I probably should find a place in the code to write that down.

Reply:
@warner PTAL with https://github.com/Agoric/agoric-sdk/pull/5252/commits/b5d29084ab7c3c7ecee762390f846e9aec6fc56e

Reply:
Yeah I'd rather find a way that doesn't require updating tags when moving from queue to queue. I see `enq` as a generic "on some kernel queue". I need to understand the use cases for an actionable queue descriptors, but processing all target, return and slots for every queue switch doesn't seem right.